src/TraceEvent/EventPipe/EventCache.cs

```diff
@@ -0,0 +1,282 @@
+                    if (droppedEvents > 0)
+                    {
+                        OnEventsDropped?.Invoke(droppedEvents);
+                    }
+                    else
+                    {
+                        // When a thread id is recycled the sequenceNumber can abruptly reset to 1 which
+                        // makes droppedEvents go negative
+                        Debug.Assert(droppedEvents == 0 || sequenceNumber == 1);
+                    }
+                    thread.SequenceNumber = sequenceNumber;
+                }
+                cursor += 12;
+            }
+        }
+
+        /// <summary>
+        /// After all events have been parsed we could have some straglers that weren't
+        /// earlier than any sorted event. Sort and dispatch those now.
+        /// </summary>
+        public void Flush()
+        {
+            SortAndDispatch(long.MaxValue);
+        }
+
+        private unsafe void SortAndDispatch(long stopTimestamp)
+        {
+            // This sort could be made faster by using a min-heap but this is a simple place to start
+            List<Queue<EventMarker>> threadQueues = new List<Queue<EventMarker>>(_threads.Values.Select(t => t.Events));
+            while(true)
+            {
+                long lowestTimestamp = stopTimestamp;
+                Queue<EventMarker> oldestEventQueue = null;
+                foreach(Queue<EventMarker> threadQueue in threadQueues)
+                {
+                    if(threadQueue.Count == 0)
+                    {
+                        continue;
+                    }
+                    long eventTimestamp = threadQueue.Peek().Header.TimeStamp;
+                    if (eventTimestamp < lowestTimestamp)
+                    {
+                        oldestEventQueue = threadQueue;
+                        lowestTimestamp = eventTimestamp;
+                    }
+                }
+                if(oldestEventQueue == null)
+                {
+                    break;
+                }
+                else
+                {
+                    EventMarker eventMarker = oldestEventQueue.Dequeue();
+                    OnEvent?.Invoke(ref eventMarker.Header);
+                }
+            }
+
+            // If the app creates and destroys threads over time we need to flush old threads
+            // from the cache or memory usage will grow unbounded. AddThread handles the
+            // the thread objects but the storage for the queue elements also does not shrink
+            // below the high water mark unless we free it explicitly. Although not unbounded
+            // growth our current runtime policy reads ahead up to 10,000 events ahead per-thread
+            // * 5000 threads * 24 bytes = ~1GB. It would be nice not to leave that much memory
+            // laying around probably mostly unused.
+            foreach(Queue<EventMarker> q in threadQueues)
```

Comment:
Don't block on this change, but we should do some testing of reasonable size profiles to make sure that we don't OOM on 32-bit PerfView.

Reply:
Are you using 'reasonable' to mean 'typical' or 'big without being crazy'? In my test scenario that does a lot of memory allocation with verbose GC + stack sampling with parallel tasks 30 sec execution creates about 5MB of data (~80,000 events). 'Typical' might be running that for a few minutes (~30MB), 'big without being crazy' might be running that 24 hours (14.4GB)?
The typical case probably won't tax the cache even if we are doing it badly because the file won't be big enough. I haven't eliminated the 32 bit position offsets in the deserializer yet so anything over 4GB will fail at this point, but we could certainly generate a 3.9GB trace.

Also thanks for reminding me about this comment because the story has actually changed a bit and I should update it. The new story is that the stream includes sequence points and nothing other than event metadata is cached across a sequence point (the reader code implements this, its just the comment that is outdated). The runtime triggers the insertion of a sequence point for every X amount of buffer space that is allocated (current 10MB, could become configurable). In the ideal world this would translate to at most 10MB of data in the stream and then 10MB of data cached in the reader. However in the hypothetical worst case the runtime could have a huge number of threads, each of those threads could have already allocated a 1MB buffer which they all fill before any of them needs to make a new allocation. Generating anything close to that pattern accidentally seems unlikely but I am happy to put that assertion to the test on some sample workloads.

Reply:
These sound like very reasonable sizes and so I'm not worried about them.  I mostly wanted to make sure that we weren't going to have to route lots of people to PerfView64.  I'm not super-concerned about pathological cases or 24 hours of trace.  In those cases with ETW, we already force folks to seek because PerfView can't handle loading traces that big into memory at once.